src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/WhenShouldBeUsedCheck.java

```diff
@@ -0,0 +1,183 @@
+ * </ul>
+ *
+ * @since 10.18.0
+ */
+
+@StatelessCheck
+public class WhenShouldBeUsedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "when.should.be.used";
+
+    /**
+     *  An array of acceptable statements inside the case block.
+     */
+    private static final int[] ACCEPTABLE_STATEMENTS_INSIDE_CASE_BLOCK = {
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_BREAK,
+        TokenTypes.EMPTY_STAT,
+        TokenTypes.RCURLY,
+    };
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LITERAL_CASE};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final boolean hasPatternLabel = hasPatternLabel(ast);
+        final DetailAST statementList = getStatementList(ast);
+        final boolean isInSwitchRule = ast.getParent().getType() == TokenTypes.SWITCH_RULE;
+
+        if (hasPatternLabel && statementList != null && isInSwitchRule) {
+            final List<DetailAST> blockStatements = getBlockStatements(statementList);
+
+            final boolean hasAcceptableStatementsOnly = blockStatements.stream()
+                    .allMatch(WhenShouldBeUsedCheck::isAcceptableStatement);
+
+            final boolean hasSingleIfWithNoElse = blockStatements.stream()
+                    .filter(WhenShouldBeUsedCheck::isSingleIfWithNoElse)
+                    .count() == 1;
+
+            if (hasAcceptableStatementsOnly && hasSingleIfWithNoElse) {
+                log(ast, MSG_KEY);
+            }
+        }
+    }
+
+    /**
+     * Get the statement list token of the case block.
+     *
+     * @param caseAST the AST node representing {@code LITERAL_CASE}
+     * @return the AST node representing {@code SLIST} of the current case
+     */
+    private static DetailAST getStatementList(DetailAST caseAST) {
+        final DetailAST caseParent = caseAST.getParent();
+        return caseParent.findFirstToken(TokenTypes.SLIST);
+    }
+
+    /**
+     * Get all statements inside the case block.
+     *
+     * @param statementList the AST node representing {@code SLIST} of the current case
+     * @return statements inside the current case block
+     */
+    private static List<DetailAST> getBlockStatements(DetailAST statementList) {
+        final List<DetailAST> blockStatements = new ArrayList<>();
+        DetailAST ast = statementList.getFirstChild();
+        while (ast != null) {
+            blockStatements.add(ast);
+            ast = ast.getNextSibling();
+        }
+        return Collections.unmodifiableList(blockStatements);
+    }
+
+    /**
+     * Check if the statement is an acceptable statement inside the case block.
```

Comment:
Please explain the concept of an acceptable statement here.

Reply:
done